### ISSUE TITLE
get rid of mask matrices and instead use `torch.index_select`

### ISSUE THREAD

**Issue by wilsonmr** [wilsonmr/anvil#1]
as above I should check if I can get rid of the matrix transformations and use torch.index_select instead, probably it's more efficient and removes multiplication by incredibly sparse tensors

**Comment by wilsonmr** [comment]
and I guess I might need to `torch.cat` the output back to a valid state
